### PR TITLE
Peerset: reuse Slots data structure for discovered peers

### PR DESCRIPTION
**Pull request by debris**
changes:
- `in_slots`, `out_slots` and `discovered` now utilize the same data structure
- discovered nodes list is now limited to `1000` entries
- `clear_common_slots` is now faster
- `add_peer` is now faster in a case where we need to swap common node with a reserved node
- `Slots` implements custom `fmt::Debug` which is more readable

### CODE REVIEW

File: core/peerset/src/slots.rs [hunk 1]
```diff
@@ -104,45 +122,88 @@ impl Slots {
...
+
+		if reserved_only {
+			return None;
+		}
+
+		self.common.pop_front()
+			.map(|(peer_id, _)| (peer_id, SlotType::Common))
+	}
+
+	pub fn clear_common_slots(&mut self) -> impl Iterator<Item = PeerId> {
```
Comment (rphmeier) [inline core/peerset/src/slots.rs]: docs on public items?

File: core/peerset/src/slots.rs [hunk 2]
```diff
@@ -62,39 +83,36 @@ impl Slots {
...
+		self.common.contains_key(peer_id) || self.reserved.contains_key(peer_id)
 	}
 
 	/// Tries to find a slot for a given peer and returns `SlotState`.
 	pub fn add_peer(&mut self, peer_id: PeerId, slot_type: SlotType) -> SlotState {
-		if let Some(st) = self.slots.get_mut(&peer_id) {
-			if *st == SlotType::Common && slot_type == SlotType::Reserved {
-				*st = SlotType::Reserved;
+		if self.reserved.contains_key(&peer_id) {
+			return SlotState::AlreadyConnected(peer_id);
```
Comment (rphmeier) [inline core/peerset/src/slots.rs]: i notice that it doesn't do a corresponding  "downgrade" of the slot type, which means that there is no invariant at the end of the function that `peer_id` is in a slot of `slot_type`. Can it have any fallout? I would lean towards documenting those behaviors

File: core/peerset/src/slots.rs [hunk 3]
```diff
@@ -104,45 +122,88 @@ impl Slots {
...
-					common_peers.push(peer_id);
-				},
-				SlotType::Reserved => {
-					self.slots.insert(peer_id, slot_type);
-				},
-			}
+	/// Pops the oldest peer from the list.
+	pub fn pop_peer(&mut self, reserved_only: bool) -> Option<(PeerId, SlotType)> {
+		if let Some((peer_id, _)) = self.reserved.pop_front() {
+			return Some((peer_id, SlotType::Reserved));
```
Comment (rphmeier) [inline core/peerset/src/slots.rs]: It seems kind of strange to pop the reserved peers first. wouldn't we rather cycle out the common ones earlier?

Comment (debris) [reply core/peerset/src/slots.rs]: I see where ur confusion comes from. `pop_peer` is only used for `discovered` list. We pop a peer that is later inserted into `out_slots` or `in_slots`

**Comment by debris** [comment]
Thanks, I will add/update the docs

**Comment by tomaka** [comment]
What's wrong with copy-pasting the code of `Slots` in a different struct?
You seem to be using some methods only in the context of connected nodes, and `pop_peer` only in the context of discovered. Half of the documentation of `Slots` mentions "we are connected to this peer", which is clearly wrong for discovered nodes.

**Comment by debris** [comment]
> What's wrong with copy-pasting the code of `Slots` in a different struct?

This is a stupid question. Why do we write functions when we can just copy-paste the code everywhere?

I realized that 6/7 methods in old `Discovered` and exactly the same as in `Slots`. 3/3 properties are the same. There is no point in copying the code.

@rphmeier I updated the docs
@tomaka I also removed references to a `connection` from `Slots` docs in every place where it was misleading

**Comment by tomaka** [comment]
In my head, you're going to have to add more fields in the future, such as "how long have we been connected to this node" or "which parachain this slot should be reserved for", and have to split out `Discovered` again.

**Comment by debris** [comment]
But we don't have those things now. And it would be premature to keep this code separate because of these reasons

**Comment by debris** [comment]
@rphmeier @tomaka can we somehow more forward with this pr?

**Pull request merged**